lib/backend/watchersyncer/watchercache.go

```diff
@@ -428,9 +430,23 @@ func (wc *watcherCache) markAsValid(resourceKey string) {
 // exceeds the error threshold.  See finishResync() for how the watcherCache goes back to in-sync.
 func (wc *watcherCache) onError() {
 	wc.errors++
-	if wc.hasSynced && wc.errors > ErrorThreshold {
-		wc.logger.WithFields(logrus.Fields{"errors": wc.errors, "threshold": ErrorThreshold}).Debugf("Exceeded error threshold")
+	threshold := GetErrorThreshold()
+	if wc.hasSynced && wc.errors > threshold {
+		wc.logger.WithFields(logrus.Fields{"errors": wc.errors, "threshold": threshold}).Debugf("Exceeded error threshold")
 		wc.hasSynced = false
 		wc.results <- api.WaitForDatastore
 	}
 }
+
+func GetErrorThreshold() int {
+	errorThresholdMtx.Lock()
+	val := errorThreshold
+	errorThresholdMtx.Unlock()
+	return val
+}
+
+func SetErrorThreshold(val int) {
```

Comment:
Having `errorThreshold` be a package-level variable makes it harder to test `watcherCache` and requires these mutex functions to prevent race conditions.

How is `watcherCache` used? Is anyone ever updating the `errorThreshold` besides right after initialization? If not, couldn't we just get rid of the mutex & global variable altogether by moving the `errorThreshold` to be a argument passed to `newWatcherCache`?

Reply:
There are those deferred goroutines in the test package that modify `errorThreshold`, besides when it is initialized. It was already a package-level variable, which seems to be necessary as it is controlling a threshold applicable to all watcherCaches.

Reply:
I think that might be better. The only time `errorThreshold` ever changes is in tests.

Reply:
@lmm Can you clarify what the idea is here.. this was already a global variable, did you mean to change that (for all consumers)?

Reply:
Yeah. I'm suggesting (and I think @lmm is agreeing) that we move errorThreshold to be a "readonly" private field on the watcherCache struct, and add it as an argument to newWatcherCache.

Reply:
>It was already a package-level variable, which seems to be necessary as it is controlling a threshold applicable to all watcherCaches.

This is the key point. I'm questioning if we really need a setting for "all watcherCaches", and if instead we just need a setting for "each watcherCache".

Reply:
> This is the key point. I'm questioning if we really need a setting for "all watcherCaches", and if instead we just need a setting for "each watcherCache".

Yup, that's what I'm trying to get at :-)

Reply:
This check in run() seems to imply it does want to look at all objects

`                                if wc.errors > GetErrorThreshold() {
                                        // Trigger a full resync if we're past the error threshold.
                                        wc.currentWatchRevision = ""
                                        wc.resyncAndCreateWatcher(ctx)
                                }
`

Reply:
No, that's just resync on that particular watcherCache. Each of the watcherCache's do their own thing.

Reply:
Ok, so you do want a per watcherCache threshold.